### PR TITLE
fix(forms): ensure model_to_dict handles empty field lists (swev-id: django__django-11163)

### LINKED ISSUE #102: model_to_dict() should return an empty dict for fields=[]
Issue summary

User request:
Been called as model_to_dict(instance, fields=[]) function should return empty dict, because no fields were requested. But it returns all fields.
Problem point is:
if fields and f.name not in fields:
which should be:
if fields is not None and f.name not in fields:

Rationale:
- Aligns with the docstring and intuitive semantics: fields=None selects all editable fields, fields=[] selects none.
- Upstream context for reference: https://github.com/django/django/pull/11150/files

Researcher specification (Emerson Gray)
- Expected behavior matrix:
  - fields=None (default): include all editable fields (subject to exclude).
  - fields=[]: return an empty dict {} regardless of exclude.
  - fields=[...] (non-empty): include only the listed editable fields.
  - exclude=None or []: no exclusion; exclude=[...]: remove listed names even if present in fields.
- Field types:
  - ForeignKey values are the raw FK column value (pk/to_field), not the related instance.
  - ManyToMany values are realized lists of related instances; [] for unsaved instance.
- Edge cases:
  - Primary key (AutoField) included by default; excluded by fields=[].
  - Non-editable fields are always excluded.
  - Behavior for deferred/private/file/choice fields is unchanged.
- Implementation change:
  - In django/forms/models.py model_to_dict(): replace `if fields and f.name not in fields:` with `if fields is not None and f.name not in fields:`
- Tests to add:
  - fields=[] yields {}.
  - fields=[] with exclude still yields {}.
  - fields subset with/without exclude interactions.
  - M2M and FK respect fields=[], and explicit subsets.
  - fields=None remains unchanged; exclude=[] is a no-op.
- Docs:
  - Clarify None vs empty list semantics in model_to_dict() docstring.
- Backwards-compat:
  - Projects that relied on fields=[] behaving like “all fields” must switch to fields=None.
  - Treat as a bug fix (see upstream PR #11150).

Proposed plan
- Implement the minimal code change described above.
- Add the tests listed to cover empty list semantics and interactions.
- Update model_to_dict docstring to clarify the semantics.

Reproduction steps (conceptual)
1) Create any model instance with editable fields.
2) Call model_to_dict(instance, fields=[])
Observed: returns a dict with all editable fields.
Expected: {}

We will attach concrete failing test output and a stack trace in the PR for traceability.

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- resolve #102 by treating an empty `fields` list as no fields while leaving `None` behavior unchanged
- document the distinction between `fields=None` and `fields=[]` and how `exclude` overrides `fields`
- add regression coverage for empty fields/exclude combinations plus foreign key and many-to-many cases

## Reproduction steps
1. create and activate a Python 3.11 virtualenv
2. install project and test requirements (`pip install -e .` and `pip install -r tests/requirements/py3.txt`)
3. run `PYTHONPATH=/workspace/django/tests:/workspace/django python tests/runtests.py model_forms --settings=test_sqlite --parallel=1`

## Observed failure stack trace
```
======================================================================
FAIL: test_empty_fields_returns_empty_dict (model_forms.tests.ModelToDictTests.test_empty_fields_returns_empty_dict)
----------------------------------------------------------------------
AssertionError: {'id': 1, 'name': 'Alice'} != {}
- {'id': 1, 'name': 'Alice'}
+ {}

======================================================================
FAIL: test_fields_empty_with_exclude_still_empty (model_forms.tests.ModelToDictTests.test_fields_empty_with_exclude_still_empty)
----------------------------------------------------------------------
AssertionError: {'id': 1} != {}
- {'id': 1}
+ {}
```

(Full failure log captured in `artifacts/failing-before.txt`.)

## Fix
- guard `model_to_dict()` with `fields is not None` so an empty iterable yields an empty mapping
- clarify the function docstring and add unit tests covering empty `fields`, `exclude` precedence, foreign keys, and many-to-many relations

## Upstream reference
- https://github.com/django/django/pull/11150/files

## Testing
- `PYTHONPATH=/workspace/django/tests:/workspace/django python tests/runtests.py model_forms --settings=test_sqlite --parallel=1`
- `flake8 django/forms/models.py`
- `flake8 --extend-ignore=E501 tests/model_forms/tests.py`

## Notes
- CI not triggered manually per instructions.

### CODE REVIEW

**Comment by casey-brooks** [comment]
### Test & Lint Summary
- `PYTHONPATH=/workspace/django/tests:/workspace/django python tests/runtests.py model_forms --settings=test_sqlite --parallel=1` → 171 passed, 0 failed, 0 skipped
- `flake8 django/forms/models.py` → no issues
- `flake8 --extend-ignore=E501 tests/model_forms/tests.py` → no issues (legacy long lines ignored)

File: artifacts/failing-before.txt [hunk 1]
```diff
@@ -0,0 +1,49 @@
+Creating test database for alias 'default'...
```
Comment (noa-lucent) [inline artifacts/failing-before.txt]: [major] Thanks for capturing the failing output, but we shouldn't commit build logs or other environment-specific artifacts to the repository. They go stale immediately and create churn for future contributors. Keeping the log in the PR description (or as an attached artifact) is enough to document the regression. Please drop \.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Thanks for digging into this regression—the logic tweak and the new regression coverage look solid. The only blocker is that we shouldn't commit build artifacts (\). Please drop that file and we should be good.

File: artifacts/failing-before.txt (hunk 1 above)
Comment (noa-lucent) [reply artifacts/failing-before.txt]: To be explicit: please remove the committed log file at artifacts/failing-before.txt from the PR.

**Review by noa-lucent — APPROVED** [review]
Thanks for addressing the feedback. The log file is gone, and the functional/test updates still look solid. ✅
